File: Practice Questions/Building Logic/CPP/BinarySearch.cpp
#include<iostream>
#include<vector>
#include<map>
using namespace std;

int BinarySearch(vector<int> &arr, int t){
    int low=0;
    int high = arr.size()-1;
    while(low<=high){
        int mid=low+(high-low)/2;//middle element
        
        if(arr[mid]==t){//if the target is in the middle
            return mid;
        }
        else if(arr[mid]<t){//if arr[mid]<t; then we change low to mid +1, then the whole thing is calculated again.
            low=mid+1;
        }
        else{//or high is changed to mid-1
            high=mid-1;
        }
    }
    return -1;
}

int main(){
    int n;
    cout<<"Enter the amount of numbers in the array:\n";
    cin>>n;
    
    cout<<"Enter the numbers in the array:\n";
    vector<int> arr;
    for(int i=0;i<n;i++){
        int x;
        cin>>x;
        arr.push_back(x);
    }
    
    int t;
    cout<<"Enter your target\n";
    cin>>t;
    
    if(BinarySearch(arr,t)==-1){
        cout<<"Element not present in array\n";
    }
    else{
        cout<<"Your element is present at index "<<BinarySearch(arr,t)<<" of the array";
    }
    
    return 0;
    
}

File: Practice Questions/Building Logic/Python/BinarySearch.py
def BinarySearch(arr,t):#very important condition: The array ,ust be sorted before binary search.
    low=arr[0]
    high=arr[len(arr)-1]
    while(low<=high):
        mid=low+(high-low)//2#remove the auto type conversion error by using // instead of /
        if(arr[mid]==t):
            return mid
        elif(arr[mid]<t):
            low=mid+1
        else:
            high=mid-1
    return -1
n=int(input("Enter the amount of numbers: "))
arr=[]
print("Enter the numbers")
for i in range (0,n):
    a=int(input())
    arr.append(a)

t=int(input("Enter the target: "))
if(BinarySearch(arr,t)==-1):
    print("The number is not present.")
else:
    print(f"the number is present in the array at index {BinarySearch(arr,t)}")
